### ISSUE TITLE
Architecture issues

### ISSUE THREAD

**Issue by minhnimble** [abdulrahim46/Nimble#9]
It was quite noticeable that there are quite a few improvements can be made on your ViewModel layer. As of now, all the network calls are executed directly in the ViewModel and there is no Network layer at all. Would you consider to implement a Network layer and put all the interceptor/authenticator you have created in each ViewModel into once place, helping the app adapt and retry the requests more scalable, leading to effortless new implementations in the future like force-upgrade, force-
logout, header adaption in each request and ssl pinning down the road?

For the data layer, I noticed that you are using `Codable` and parse responses with `jsonapi` format manually instead of using a 3rd-party library for handling the task. Do you think using a helper library would add lot of benefit? Why do you think parsing manually would be a better choice for the app?

This one is more of a nitpick, but I don't see you apply Dependency Injection in the project at all. Do you think it would be better to apply Dependency Injection for this project? If yes, where are the most suitable places you think that it is worth applying?

**Comment by abdulrahim46** [comment]
> It was quite noticeable that there are quite a few improvements can be made on your ViewModel layer. As of now, all the network calls are executed directly in the ViewModel and there is no Network layer at all. Would you consider to implement a Network layer and put all the interceptor/authenticator you have created in each ViewModel into once place, helping the app adapt and retry the requests with more scalability, leading to effortless new extensions in the future like force-upgrade, force-logout, header adaption and ssl pinning down the road in each request?

Yeah, we need to improve the architecture flow by adding Network/APIService layer so that we inject(using DI) that in viewmodel. So that it will be easy for us scale the network stuffs properly for future tasks.

> For the data layer, I noticed that you are using Codable and parse responses with JSON API format manually instead of using a 3rd-party library for handling the task. Why do you think parsing manually would be a better choice for the application? And do you think using a helper library would add more benefit?

This question is a difficult one to answer for a few reasons mainly because it comes down to personal preference and what is the best tool for the job you are doing.

I used to use SwiftyJSON/ObjectMapper, but after Codeable came out, I made the switch mainly because it was one less dependency for my projects. Swift keeps it up to date and working with latest features. With a third-party framework, you have to wait for them to be updated by the owners.

Personally, I would go with Codeable it's fast and reliable.

> Codable eliminates the need for any initializing function, given that we are content with our object’s properties having the same name as their JSON keys, and can assume that the JSON data is in the same format we expect the Object’s properties to be in (i.e. strings properly capitalized, URLs with or without schemes).

**Comment by abdulrahim46** [comment]
> This one is more of a nitpick, but I don't see you apply Dependency Injection in the project at all. Do you think it would be better to apply Dependency Injection for this project? If yes, where are the most suitable places you think it would be worth applying?

Yeah, It would be worth it, as mentoned in my previous comment we can inject apiServices in viewModel classes and ViewControllers for extra buiness logic that we need to implement. It also help in unit testing the ViewModels. What do you think?

> When declaring constants, for example the Constants.swift file, it is recognized that you are using struct with static. What do you think about using enum for constants? Which do you think is better between struct and enum? and why?

Both structs and enumerations work.

> The advantage of using a case-less enumeration is that it can't accidentally be instantiated and works as a pure namespace.

So for a structure,

`let foo = ErrorMessage()`

`creates a (useless) value of type ErrorMessage, but for a case-less enumeration it fails to compile:`

`let foo = ErrorMessage()`
`// error: 'ErrorMessage' cannot be constructed because it has no accessible initializers`

Do we need our constants to be unique?  Then we can use an enum, which enforces this.
If we want to use several different constants to contain the same value (often useful for clarity)? Then we can use a struct, which allows this. 
What do you think?

**Comment by abdulrahim46** [comment]
> There are a mixed of logic and UI setup in your ViewControllers. For instance, you define checkValidation with some email and password validation logic inside ViewController to decide which alert to show. What is your opinion on moving these logics into the ViewModel layer? And how would you extract these logic to ViewModel if it is recommended? 

Yep, you're right I'm sorry for that. This defies MVVM pattern. We can move these validation or other buisness related logic to ViewModel. For example, we can check regex or these validation in ViewModels before hitting the api itself & return an error to user or move ahead.

**Comment by minhnimble** [comment]
> This question is a difficult one to answer for a few reasons mainly because it comes down to personal preference and what is the best tool for the job you are doing.
I used to use SwiftyJSON/ObjectMapper, but after Codeable came out, I made the switch mainly because it was one less dependency for my projects. Swift keeps it up to date and working with latest features. With a third-party framework, you have to wait for them to be updated by the owners.
Personally, I would go with Codeable it's fast and reliable.

I think you misunderstand my question a bit, since my point is mainly about handling `JSON:API` response format. In case you don't know what it is and why we use it, please feel free to take a look [here](https://jsonapi.org) 🙏. 
At the same time, our provided documentations suggest some iOS libraries for handling `JSON:API` response [here](https://jsonapi.org/implementations/#client-libraries-ios) and you probably missed that. 😅
Anyway, what do you think about updating to use `JSON:API` parser instead of doing it manually with `Codable` like now?

> So for a structure, 
`let foo = ErrorMessage()`
creates a (useless) value of type ErrorMessage, but for a case-less enumeration it fails to compile:
`let foo = ErrorMessage()`
// error: 'ErrorMessage' cannot be constructed because it has no accessible initializers
Do we need our constants to be unique? Then we can use an enum, which enforces this.
If we want to use several different constants to contain the same value (often useful for clarity)? Then we can use a struct, which allows this.
What do you think?

It is great to see that you know when to use a `struct` or `enum`. Now from that understanding I don't see you made the finalized decision to go with `struct` or `enum` yet. Does it mean that you still want to stick with the current implementation or update to use `enum? I am all ears for your final decision here 🙏

**Comment by abdulrahim46** [comment]
> I think you misunderstand my question a bit, since my point is mainly about handling JSON:API response format. In case you don't know what it is and why we use it, please feel free to take a look here 🙏.
At the same time, our provided documentations suggest some iOS libraries for handling JSON:API response here and you probably missed that. 😅
Anyway, what do you think about updating to use a JSON:API parser instead of doing it manually with Codable like now?

Ohh got it. Sorry for misunderstanding;😅 Yeah I saw the libraries that mentioned in the [documentation](https://jsonapi.org/implementations/#client-libraries-ios) but I decided to do it manually because of the extra dependency. So, Do you want me to integrate the library for this also?

**Comment by minhnimble** [comment]
> Ohh got it. Sorry for misunderstanding;😅 Yeah I saw the libraries that mentioned in the documentation but I decided to do it manually because of the extra dependency. So, Do you want me to integrate the library for this also?

@abdulrahim46 yes, it would be nice if to see how you implement the library for `JSON:API` 🙏
